core/iwasm/common/wasm_runtime_common.c

```diff
@@ -3012,8 +3012,74 @@ wasm_runtime_init_wasi(WASMModuleInstanceCommon *module_inst,
             goto fail;
         }
 
-        fd_table_insert_existing(curfds, wasm_fd, raw_fd);
-        fd_prestats_insert(prestats, dir_list[i], wasm_fd);
+        if (!fd_table_insert_existing(curfds, wasm_fd, raw_fd)
+            || !fd_prestats_insert(prestats, dir_list[i], wasm_fd)) {
+            if (error_buf)
+                snprintf(
+                    error_buf, error_buf_size,
+                    "error while pre-opening directory %s: insertion failed\n",
+                    dir_list[i]);
+            goto fail;
+        }
+    }
+
+    for (i = 0; i < map_dir_count; i++, wasm_fd++) {
+        char mapping_copy_buf[256];
+        char *mapping_copy = mapping_copy_buf;
+        char *map_mapped = NULL, *map_host = NULL;
+        const unsigned long max_len = strlen(map_dir_list[i]) * 2 + 3;
+
+        // Allocation limit for runtime environments with reduced stack size
+        if (max_len > 256) {
+            if (!(mapping_copy = wasm_runtime_malloc(max_len))) {
+                snprintf(error_buf, error_buf_size,
+                         "error while allocating for directory mapping\n");
+                goto fail;
+            }
+        }
+
+        strncpy(mapping_copy, map_dir_list[i], strlen(map_dir_list[i]));
```

Comment:
The mapping_copy may be not null-terminated after copy, see manual of strncpy:
`Warning: If there is no null byte among the first n  bytes
       of src, the string placed in dest will not be null-terminated.`

Had better strcpy or `strncpy(mapping_copy, map_dir_list[i], strlen(map_dir_list[i]) + 1)`?

Reply:
Thanks for correcting, done.